drivers/pcie/endpoint/pcie_ep_common.c

```diff
@@ -90,21 +142,149 @@ int pcie_ep_xfer_data_memcpy(const struct device *dev, uint64_t pcie_addr,
 		return mapped_size;
 	}
 
+	/*
+	 * In second attempt, we must have mapped entire size,
+	 * if not just quit here before attempting memcpy
+	 */
+	if (mapped_size != unmapped_size) {
+		pcie_ep_unmap_addr(dev, mapped_addr);
+		return -EIO;
+	}
+
 	ret = pcie_ep_mapped_copy(mapped_addr,
 				  ((uintptr_t)local_addr) + xfer_size,
 				  mapped_size, dir);
 
+	/* Flush the PCIe writes upon successful memcpy */
+	if (!ret && (dir == DEVICE_TO_HOST)) {
+		sys_read8(mapped_addr);
+	}
+
 	pcie_ep_unmap_addr(dev, mapped_addr);
 
 	/* Check if mapped_copy succeeded */
 	if (ret < 0) {
 		return ret;
 	}
 
-	/* In second attempt, we must have completed data transfer */
+	return 0;
+}
+
+/*
+ * Helper API to achieve data transfer with DMA operation through
+ * PCIe outbound memory, this API is based off pcie_ep_xfer_data_memcpy,
+ * here we use "system dma" instead of memcpy
+ */
+int pcie_ep_xfer_data_dma(const struct device *dev, uint64_t pcie_addr,
+			  uintptr_t *local_addr, uint32_t size,
+			  enum pcie_ob_mem_type ob_mem_type,
+			  enum xfer_direction dir)
+{
+	uint64_t mapped_addr;
+	int mapped_size, ret;
+	uint32_t xfer_size, unmapped_size;
+	uint32_t dummy_data;	/* For explicit dummy PCIe read */
+
+	mapped_size = pcie_ep_map_addr(dev, pcie_addr, &mapped_addr,
+				       size, ob_mem_type);
+
+	/* Check if outbound memory mapping succeeded */
+	if (mapped_size < 0) {
+		return mapped_size;
+	}
+
+	ret = pcie_ep_dma_xfer(dev, mapped_addr, (uintptr_t)local_addr,
+			       mapped_size, dir);
+
+	/* Flush the PCIe writes upon successful DMA */
+	if (!ret && (dir == DEVICE_TO_HOST)) {
+		if (IS_ENABLED(CONFIG_64BIT) || !(mapped_addr >> 32)) {
```

Comment:
this could be checked in previous if () already via an && ()

In fact, put the IS_ENABLED(CONFIG_64BIT) first tested, since the whole logic here is meant only for 64 bit.

Reply:
> In fact, put the IS_ENABLED(CONFIG_64BIT) first tested, since the whole logic here is meant only for 64 bit.

I do not understand this comment. `sys_read8` is not meant only for 64-bit system, it is meant for either 64-bit system or 32-bit system, doing dma with lowmem. (above table in the comment can be referred for clarity)

> this could be checked in previous if () already via an && ()

This can be done, but current way seems more readable, i.e. first line to indicate this is needed only for PCIe writes (`DEVICE_TO_HOST`) and second line to check if 64 bit system(`CONFIG_64BIT`) or 32-bit system using lowmem(`!(mapped_addr >> 32)`).

Reply:
Oh yes I miss-read the '||', nevermind